File: src/EasyPrinter/Services/ScannerService.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Threading.Tasks;
using EasyPrinter.Models;
using NAPS2.Scan;
using PdfSharp.Drawing;
using PdfSharp.Pdf;

namespace EasyPrinter.Services
{
    /// <summary>
    /// Сервис сканирования документов через WIA
    /// </summary>
    public class ScannerService : IDisposable
    {
        private readonly ImageProcessingService _imageProcessing;
        private bool _isDisposed;

        public event EventHandler<ScanProgressEventArgs>? ScanProgress;
        public event EventHandler<ScanCompletedEventArgs>? ScanCompleted;

        public ScannerService(ImageProcessingService imageProcessing)
        {
            _imageProcessing = imageProcessing;
        }

        /// <summary>
        /// Выполнить сканирование
        /// </summary>
        public async Task<Image?> ScanAsync(ScanSettings settings)
        {
            try
            {
                ScanProgress?.Invoke(this, new ScanProgressEventArgs { Message = "Поиск сканера...", Progress = 10 });

                using var scanningContext = new ScanningContext();

                // Получаем список доступных сканеров
                var controller = new ScanController(scanningContext);
                var devices = await controller.GetDeviceList();

                // Ищем HP сканер
                ScanDevice? hpScanner = null;
                foreach (var device in devices)
                {
                    if (device.Name.Contains("HP", StringComparison.OrdinalIgnoreCase) ||
                        device.Name.Contains("1536", StringComparison.OrdinalIgnoreCase))
                    {
                        hpScanner = device;
                        break;
                    }
                }

                if (hpScanner == null && devices.Count > 0)
                {
                    // Берём первый доступный сканер
                    hpScanner = devices[0];
                }

                if (hpScanner == null)
                {
                    throw new InvalidOperationException("Сканер не найден");
                }

                ScanProgress?.Invoke(this, new ScanProgressEventArgs
                {
                    Message = $"Сканирование на {hpScanner.Name}...",
                    Progress = 30
                });

                // Настройки сканирования
                var scanOptions = new ScanOptions
                {
                    Device = hpScanner,
                    Dpi = (int)settings.Resolution,
                    PaperSource = settings.Source == ScanSource.ADF
                        ? NAPS2.Scan.PaperSource.Feeder
                        : NAPS2.Scan.PaperSource.Flatbed,
                    BitDepth = BitDepth.Color,
                    PageSize = PageSize.A4
                };

                // Выполняем сканирование
                Image? resultImage = null;

                await foreach (var image in controller.Scan(scanOptions))
                {
                    ScanProgress?.Invoke(this, new ScanProgressEventArgs
                    {
                        Message = "Обработка изображения...",
                        Progress = 70
                    });

                    // Конвертируем NAPS2 изображение в System.Drawing.Image
                    using var stream = new MemoryStream();
                    image.Save(stream, ImageFileFormat.Png);
                    stream.Position = 0;
                    resultImage = Image.FromStream(stream);

                    // Применяем настройки изображения если есть
                    if (settings.ImageAdjustments.HasChanges)
                    {
                        var processed = _imageProcessing.ApplyAdjustments(resultImage, settings.ImageAdjustments);
                        resultImage.Dispose();
                        resultImage = processed;
                    }

                    // Для простоты берём только первую страницу
                    // Для многостраничного сканирования нужна дополнительная логика
                    break;
                }

                ScanProgress?.Invoke(this, new ScanProgressEventArgs
                {
                    Message = "Сканирование завершено",
                    Progress = 100
                });

                return resultImage;
            }
            catch (Exception ex)
            {
                ScanCompleted?.Invoke(this, new ScanCompletedEventArgs
                {
                    Success = false,
                    Error = ex.Message
                });
                throw;
            }
        }

        /// <summary>
        /// Сохранить отсканированное изображение
        /// </summary>
        public void SaveScan(Image image, ScanSettings settings)
        {
            var outputPath = settings.GetFullPath();

            // Создаём директорию если не существует
            var directory = Path.GetDirectoryName(outputPath);
            if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
            {
                Directory.CreateDirectory(directory);
            }

            switch (settings.Format)
            {
                case ScanFormat.PDF:
                    SaveAsPdf(image, outputPath);
                    break;
                case ScanFormat.JPEG:
                    image.Save(outputPath, ImageFormat.Jpeg);
                    break;
                case ScanFormat.PNG:
                    image.Save(outputPath, ImageFormat.Png);
                    break;
                case ScanFormat.TIFF:
                    image.Save(outputPath, ImageFormat.Tiff);
                    break;
            }

            ScanCompleted?.Invoke(this, new ScanCompletedEventArgs
            {
                Success = true,
                FilePath = outputPath
            });
        }

        /// <summary>
        /// Сохранить изображение как PDF
        /// </summary>
        private void SaveAsPdf(Image image, string outputPath)
        {
            using var document = new PdfDocument();
            var page = document.AddPage();

            // Устанавливаем размер страницы по размеру изображения
            page.Width = XUnit.FromPoint(image.Width * 72 / image.HorizontalResolution);
            page.Height = XUnit.FromPoint(image.Height * 72 / image.VerticalResolution);

            using var graphics = XGraphics.FromPdfPage(page);

            // Сохраняем изображение во временный файл для XImage
            var tempPath = Path.GetTempFileName() + ".png";
            try
            {
                image.Save(tempPath, ImageFormat.Png);
                using var xImage = XImage.FromFile(tempPath);
                graphics.DrawImage(xImage, 0, 0, page.Width, page.Height);
            }
            finally
            {
                if (File.Exists(tempPath))
                    File.Delete(tempPath);
            }

            document.Save(outputPath);
        }

        /// <summary>
        /// Получить список доступных сканеров
        /// </summary>
        public async Task<string[]> GetAvailableScannersAsync()
        {
            try
            {
                using var scanningContext = new ScanningContext();
                var controller = new ScanController(scanningContext);
                var devices = await controller.GetDeviceList();

                var names = new string[devices.Count];
                for (int i = 0; i < devices.Count; i++)
                {
                    names[i] = devices[i].Name;
                }
                return names;
            }
            catch
            {
                return Array.Empty<string>();
            }
        }

        public void Dispose()
        {
            if (!_isDisposed)
            {
                _isDisposed = true;
            }
        }
    }

    public class ScanProgressEventArgs : EventArgs
    {
        public string Message { get; set; } = string.Empty;
        public int Progress { get; set; }
    }

    public class ScanCompletedEventArgs : EventArgs
    {
        public bool Success { get; set; }
        public string? FilePath { get; set; }
        public string? Error { get; set; }
    }
}

File: easyprinter/services/scanner_service.py
"""
Сервис сканирования документов
"""

import os
import platform
import subprocess
import tempfile
from datetime import datetime
from typing import Optional, Callable, List
from PIL import Image
from io import BytesIO

from ..models import ScanSettings, ScanFormat, ScanSource, ScanResolution
from .image_processing_service import ImageProcessingService


class ScanProgressEvent:
    """Событие прогресса сканирования"""

    def __init__(self, message: str = "", progress: int = 0):
        self.message = message
        self.progress = progress


class ScanCompletedEvent:
    """Событие завершения сканирования"""

    def __init__(self, success: bool = False, file_path: Optional[str] = None, error: Optional[str] = None):
        self.success = success
        self.file_path = file_path
        self.error = error


class ScannerService:
    """Сервис сканирования документов"""

    def __init__(self, image_processing: ImageProcessingService):
        self._image_processing = image_processing
        self._is_disposed = False
        self._progress_callbacks: List[Callable[[ScanProgressEvent], None]] = []
        self._completed_callbacks: List[Callable[[ScanCompletedEvent], None]] = []

    def add_progress_callback(self, callback: Callable[[ScanProgressEvent], None]) -> None:
        """Добавить callback для события прогресса"""
        self._progress_callbacks.append(callback)

    def add_completed_callback(self, callback: Callable[[ScanCompletedEvent], None]) -> None:
        """Добавить callback для события завершения"""
        self._completed_callbacks.append(callback)

    def _notify_progress(self, message: str, progress: int) -> None:
        """Уведомить о прогрессе"""
        event = ScanProgressEvent(message, progress)
        for callback in self._progress_callbacks:
            try:
                callback(event)
            except Exception:
                pass

    def _notify_completed(self, success: bool, file_path: Optional[str] = None, error: Optional[str] = None) -> None:
        """Уведомить о завершении"""
        event = ScanCompletedEvent(success, file_path, error)
        for callback in self._completed_callbacks:
            try:
                callback(event)
            except Exception:
                pass

    def scan(self, settings: ScanSettings) -> Optional[Image.Image]:
        """Выполнить сканирование"""
        try:
            self._notify_progress("Поиск сканера...", 10)

            system = platform.system()

            if system == "Windows":
                image = self._scan_windows(settings)
            elif system == "Darwin":
                image = self._scan_macos(settings)
            else:
                image = self._scan_linux(settings)

            if image is None:
                raise RuntimeError("Не удалось получить изображение от сканера")

            self._notify_progress("Обработка изображения...", 70)

            # Применяем настройки изображения если есть
            if settings.image_adjustments.has_changes:
                image = self._image_processing.apply_adjustments(image, settings.image_adjustments)

            self._notify_progress("Сканирование завершено", 100)

            return image

        except Exception as e:
            self._notify_completed(False, error=str(e))
            raise

    def _scan_windows(self, settings: ScanSettings) -> Optional[Image.Image]:
        """Сканирование на Windows через WIA"""
        self._notify_progress("Запуск сканирования (Windows WIA)...", 30)

        # Создаём временный файл для сохранения скана
        with tempfile.NamedTemporaryFile(suffix='.png', delete=False) as tmp:
            tmp_path = tmp.name

        try:
            # PowerShell скрипт для сканирования через WIA
            dpi = settings.resolution.value
            ps_script = f'''
Add-Type -AssemblyName System.Drawing

$deviceManager = New-Object -ComObject WIA.DeviceManager
$device = $null

foreach ($d in $deviceManager.DeviceInfos) {{
    if ($d.Type -eq 1) {{  # Scanner
        $device = $d.Connect()
        break
    }}
}}

if ($device -eq $null) {{
    Write-Error "Сканер не найден"
    exit 1
}}

$item = $device.Items[1]

# Настройки разрешения
$item.Properties["6146"].Value = 2  # Color
$item.Properties["6147"].Value = {dpi}  # Horizontal DPI
$item.Properties["6148"].Value = {dpi}  # Vertical DPI

# Сканирование
$imageFile = $item.Transfer()
$imageFile.SaveFile("{tmp_path.replace(chr(92), chr(92)+chr(92))}")
'''

            result = subprocess.run(
                ["powershell", "-Command", ps_script],
                capture_output=True,
                text=True,
                timeout=120
            )

            if result.returncode != 0 or not os.path.exists(tmp_path):
                # Если WIA не сработал, пробуем через wiaaut.dll
                return self._scan_windows_fallback(settings)

            return Image.open(tmp_path)

        finally:
            try:
                if os.path.exists(tmp_path):
                    os.unlink(tmp_path)
            except Exception:
                pass

    def _scan_windows_fallback(self, settings: ScanSettings) -> Optional[Image.Image]:
        """Запасной метод сканирования на Windows"""
        # Открываем стандартное приложение сканирования Windows
        subprocess.run(["start", "ms-screenclip:"], shell=True, timeout=5)
        raise RuntimeError("Автоматическое сканирование недоступно. Используйте стандартное приложение Windows для сканирования.")

    def _scan_macos(self, settings: ScanSettings) -> Optional[Image.Image]:
        """Сканирование на macOS через Image Capture / SANE"""
        self._notify_progress("Запуск сканирования (macOS)...", 30)

        # Используем sane-backends если установлен
        with tempfile.NamedTemporaryFile(suffix='.png', delete=False) as tmp:
            tmp_path = tmp.name

        try:
            dpi = settings.resolution.value
            mode = "Color"

            # Пробуем scanimage (SANE)
            args = [
                "scanimage",
                f"--resolution={dpi}",
                f"--mode={mode}",
                f"--format=png",
                f"-o", tmp_path
            ]

            if settings.source == ScanSource.ADF:
                args.append("--source=ADF")

            result = subprocess.run(args, capture_output=True, text=True, timeout=120)

            if result.returncode == 0 and os.path.exists(tmp_path):
                return Image.open(tmp_path)
            else:
                # Открываем Image Capture
                subprocess.run(["open", "-a", "Image Capture"], timeout=5)
                raise RuntimeError("Автоматическое сканирование недоступно. Используйте Image Capture для сканирования.")

        finally:
            try:
                if os.path.exists(tmp_path):
                    os.unlink(tmp_path)
            except Exception:
                pass

    def _scan_linux(self, settings: ScanSettings) -> Optional[Image.Image]:
        """Сканирование на Linux через SANE"""
        self._notify_progress("Запуск сканирования (Linux SANE)...", 30)

        with tempfile.NamedTemporaryFile(suffix='.png', delete=False) as tmp:
            tmp_path = tmp.name

        try:
            dpi = settings.resolution.value

            args = [
                "scanimage",
                f"--resolution={dpi}",
                "--mode=Color",
                "--format=png",
                "-o", tmp_path
            ]

            if settings.source == ScanSource.ADF:
                args.append("--source=ADF")

            result = subprocess.run(args, capture_output=True, text=True, timeout=120)

            if result.returncode == 0 and os.path.exists(tmp_path):
                return Image.open(tmp_path)
            else:
                raise RuntimeError(f"Ошибка сканирования: {result.stderr}")

        finally:
            try:
                if os.path.exists(tmp_path):
                    os.unlink(tmp_path)
            except Exception:
                pass

    def save_scan(self, image: Image.Image, settings: ScanSettings) -> str:
        """Сохранить отсканированное изображение"""
        output_path = settings.get_full_path()

        # Создаём директорию если не существует
        directory = os.path.dirname(output_path)
        if directory and not os.path.exists(directory):
            os.makedirs(directory)

        if settings.format == ScanFormat.PDF:
            self._save_as_pdf(image, output_path)
        elif settings.format == ScanFormat.JPEG:
            # Конвертируем в RGB если есть альфа-канал
            if image.mode == 'RGBA':
                image = image.convert('RGB')
            image.save(output_path, 'JPEG', quality=95)
        elif settings.format == ScanFormat.PNG:
            image.save(output_path, 'PNG')
        elif settings.format == ScanFormat.TIFF:
            image.save(output_path, 'TIFF')

        self._notify_completed(True, output_path)
        return output_path

    def _save_as_pdf(self, image: Image.Image, output_path: str) -> None:
        """Сохранить изображение как PDF"""
        # Конвертируем в RGB если нужно
        if image.mode == 'RGBA':
            rgb_image = Image.new('RGB', image.size, (255, 255, 255))
            rgb_image.paste(image, mask=image.split()[3])
            image = rgb_image
        elif image.mode != 'RGB':
            image = image.convert('RGB')

        image.save(output_path, 'PDF', resolution=100.0)

    def get_available_scanners(self) -> List[str]:
        """Получить список доступных сканеров"""
        system = platform.system()
        scanners = []

        try:
            if system == "Windows":
                # Используем WIA для получения списка сканеров
                ps_script = '''
$deviceManager = New-Object -ComObject WIA.DeviceManager
foreach ($d in $deviceManager.DeviceInfos) {
    if ($d.Type -eq 1) {
        Write-Output $d.Properties["Name"].Value
    }
}
'''
                result = subprocess.run(
                    ["powershell", "-Command", ps_script],
                    capture_output=True,
                    text=True,
                    timeout=10
                )
                if result.returncode == 0:
                    scanners = [s.strip() for s in result.stdout.strip().split('\n') if s.strip()]

            elif system in ("Darwin", "Linux"):
                # Используем SANE
                result = subprocess.run(
                    ["scanimage", "-L"],
                    capture_output=True,
                    text=True,
                    timeout=10
                )
                if result.returncode == 0:
                    for line in result.stdout.strip().split('\n'):
                        if line.strip():
                            # Формат: device `name' is a Vendor Model Type
                            if '`' in line and "'" in line:
                                start = line.index('`') + 1
                                end = line.index("'")
                                scanners.append(line[start:end])

        except Exception:
            pass

        return scanners

    def dispose(self) -> None:
        """Освободить ресурсы"""
        self._is_disposed = True
        self._progress_callbacks.clear()
        self._completed_callbacks.clear()
